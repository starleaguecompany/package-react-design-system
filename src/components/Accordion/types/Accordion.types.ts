import * as React from 'react'
import { Omit, tuple } from '@starleaguecompany/package-react-utils'

const AccordionSeparators = tuple('divider', 'space')
const AccordionBackgrounds = tuple('white')

export type AccordionSeparator = typeof AccordionSeparators[number]
export type AccordionBackground = typeof AccordionBackgrounds[number]

export interface AccordionProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
  /** Current Items indexes */
  defaultIndex?: Array<number>
  /** The accordion item's separator */
  separator?: AccordionSeparator
  /** The accordion item's background */
  background?: AccordionBackground
  /** Defines the behavior of the scroll. By default is true, when opening an item, scrolls the screen so that the element is visible to the user */
  scrollIntoView?: boolean
  /** Callback executed when active item is changed */
  onChange?: (indexes: Array<number>) => void
}

export interface AccordionItemProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'title'> {
  /** Active state */
  active?: boolean
  /** The accordion item's title */
  title?: React.ReactNode
  /** The accordion item's subtitle */
  subtitle?: string
  /** The accordion item's Icon */
  icon?: React.ReactNode
  /** Defines the behavior of the scroll. Should be set in Accordion props */
  scrollIntoView?: boolean
  /** Callback executed when active item is changed */
  onCollapse?: (event: React.MouseEvent<HTMLDivElement> | React.KeyboardEvent<HTMLDivElement>) => void
}
