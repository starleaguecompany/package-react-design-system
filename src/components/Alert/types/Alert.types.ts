import * as React from 'react'
import { Omit, tuple } from '@starleaguecompany/package-react-utils'

import { COLORS } from '../../../constants/colors'

const AlertColors = tuple('light', 'dark', COLORS.GREEN, COLORS.ORANGE, COLORS.RED, COLORS.BLUE)

export type AlertColor = typeof AlertColors[number]

export interface AlertProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'title'> {
  /** The alert's title */
  title?: React.ReactNode
  /** The alert's subtitle */
  subtitle?: React.ReactNode
  /** The alert's Icon */
  icon?: React.ReactNode
  /** Whether the Alert can be closed */
  closable?: boolean
  /** Color */
  color?: AlertColor
  /** Footer content */
  footer?: React.ReactNode
  /** Callback executed when tag is closed */
  onClose?: React.MouseEventHandler<HTMLSpanElement>
}
