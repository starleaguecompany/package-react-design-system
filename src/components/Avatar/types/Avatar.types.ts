import * as React from 'react';

import { CONTAINER_SIZES } from '../../../constants/sizes';

export interface AvatarProps extends React.HTMLAttributes<HTMLDivElement> {
  /** The address of the image for an image avatar */
  src?: string;
  /** The size of the avatar */
  size?: typeof CONTAINER_SIZES[keyof typeof CONTAINER_SIZES];
}
