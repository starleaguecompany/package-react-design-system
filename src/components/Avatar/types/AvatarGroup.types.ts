import * as React from 'react';

import { CONTAINER_SIZES } from '../../../constants/sizes';

export interface AvatarGroupProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Size */
  size?: typeof CONTAINER_SIZES[keyof typeof CONTAINER_SIZES];
}
