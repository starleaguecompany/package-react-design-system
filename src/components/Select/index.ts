import Select from './src/Select'
import { SelectProps, SelectValue, SelectOption } from './types/Select.types'

export { Select, SelectProps, SelectValue, SelectOption }
