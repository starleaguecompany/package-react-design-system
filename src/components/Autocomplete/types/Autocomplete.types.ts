import { Omit } from '@starleaguecompany/package-react-utils'
import * as React from 'react'

import { SelectOption } from '../../../types/Select.types'
import { FieldProps } from '../../../types/Field.types'
import { Placement } from '../../../types/Placements.types'
import { InputPlacemenInGroup } from '../../../types/InputPlacemenInGroup'

export { SelectOption as AutocompleteOption }

export type AutocompleteValue = string | number | undefined

export interface AutocompleteProps
  extends Omit<FieldProps<HTMLTextAreaElement, AutocompleteValue>, 'onChange'>,
    InputPlacemenInGroup {
  /**
   * **Deprecated:** The input postfix
   * @deprecated Since version 0.2.24 Will be removed soon
   */
  postfix?: string | undefined
  /** Select options. Will get better perf than jsx definition */
  options: Array<SelectOption>
  /** Dropdown menu position */
  placement?: Placement
  /** Fixed dropdown menu width. If `true`, the dropdown menu will match the width of the input at all times */
  fixedWidth?: boolean
  /** Allows to replace default message with custom component when no result matches */
  notFoundContent?: React.ReactNode
  /** Autocomplete height adjusts to label content. **NOTE, should be used with mobile devices only** */
  resizable?: boolean
  /** Defines the behavior of the scroll. By default is `true`, when opening a list of options, scrolls it so that the active option is visible to the user */
  scrollIntoView?: boolean
  /** Called when searching items */
  onSearch?: (value: string) => void
  /** Callback function that is fired when the user changes the select value */
  onChange?: (value: AutocompleteValue) => void
}
