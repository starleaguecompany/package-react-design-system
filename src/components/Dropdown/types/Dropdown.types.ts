import { Omit } from '@starleaguecompany/package-react-utils';

import { SelectOption, SelectValue, SelectMode } from '../../../types/Select.types';
import { Placement } from '../../../types/Placements.types';

export interface DropdownProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'defaultValue' | 'onChange'> {
  /** The select default value */
  defaultValue?: SelectValue;
  /** The select value */
  value?: SelectValue;
  /** Placement position */
  placement?: Placement;
  /** Dropdown options. Will get better perf than jsx definition */
  options: Array<SelectOption>;
  /** Multiple mode */
  mode?: SelectMode;
  /** Change callback */
  onChange?: (value: SelectValue) => void;
  /** Callback executed when dropdown is opening */
  onOpen?: () => void;
  /** Callback executed when dropdown is closing */
  onClose?: () => void;
}
