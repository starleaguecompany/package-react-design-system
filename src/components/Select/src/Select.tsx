import * as React from 'react'
import {
  useStyles,
  useStateCallback,
  useBoolean,
  useKeyPress,
  usePositioner,
  useFormControlContext,
  noop,
  safeInvoke,
  getNextIndex,
  getPrevIndex,
  useMergedRef,
  isNull,
} from '@starleaguecompany/package-react-utils'
import { ArrowDown, Cross } from '@starleaguecompany/react-icons'

import { Z_INDEXES } from '../../../constants/zIndexes'

import { Portal } from '../../Portal'
import { TextInput } from '../../TextInput'
import { Menu } from '../../Menu'
import { Fade } from '../../Transition'

import { SelectOption, SelectValue } from '../../../types/Select.types'
import { SelectProps } from '../types/Select.types'
import styles from '../styles/Select.module.less'

/**
 * @description Select component to select value from options
 *
 * @component
 * @example
 * ```jsx
 * <Select placeholder="Услуги" defaultValue={value} options={options} />
 * ```
 */
const Select = React.forwardRef<HTMLInputElement, SelectProps>((props, ref) => {
  const {
    autoFocus,
    defaultValue,
    value,
    mode,
    options,
    invalid,
    disabled,
    readOnly,
    icon,
    placement,
    flip,
    fixedWidth,
    className,
    withCrossIcon,
    onChange,
    onFocus,
    onClick,
    onBlur,
    ...restProps
  } = props

  const cx = useStyles(styles)
  const inputRef = React.useRef<HTMLInputElement>(null)
  const reference = React.useRef<Element | null>(null)
  const popper = React.useRef<HTMLElement | null>(null)
  const focusedItemRef = React.useRef<HTMLDivElement | null>(null)
  const [visible, setVisible] = useBoolean(false)
  const formControlContext = useFormControlContext()
  const [innerValue, setInnerValue] = useStateCallback<SelectValue | undefined>(value || defaultValue)
  const [focused, setFocused] = useBoolean(autoFocus || false)
  const [innerOptions, setInnerOptions] = React.useState<SelectOption[]>(options)
  const { referenceRef, popperRef, getReferenceProps, getPopperProps } = usePositioner({
    placement: placement ? placement : 'bottom-start',
    matchWidth: fixedWidth,
    offset: [0, 8],
    flip,
  })
  const [focusedIndex, setFocusedIndex] = React.useState(-1)
  const classNames = cx(className, 'container')
  const isMultiple = mode === 'multiple'

  const optionsMap = React.useMemo<Record<SelectOption['value'], SelectOption>>(() => {
    if (!innerOptions) {
      return {}
    }

    return innerOptions.reduce<Record<SelectOption['value'], SelectOption>>((acc, o) => {
      acc[o.value] = o

      return acc
    }, {})
  }, [innerOptions])

  const activeOptions = React.useMemo<SelectOption[]>(() => {
    if (!innerOptions) {
      return []
    }

    return innerOptions.filter(option => !option.disabled)
  }, [innerOptions])

  const getCurrentIndex = (value: SelectValue | undefined) => {
    if (!Array.isArray(innerOptions) || !value) return -1
    if (isMultiple && Array.isArray(value)) {
      return activeOptions.findIndex(option => option.value === value[value.length - 1])
    }

    return activeOptions.findIndex(option => option.value === value)
  }

  const displayValue = React.useMemo(() => {
    if (innerValue === undefined || innerValue === null) {
      return undefined
    }

    if (isMultiple && Array.isArray(innerValue)) {
      if (innerValue.length === 0) {
        return undefined
      }

      if (innerOptions) {
        return innerValue
          .map(v => optionsMap[v]?.label)
          .filter(Boolean)
          .join(', ')
      }
    }

    return optionsMap[innerValue as string]?.label
  }, [innerValue, optionsMap])

  const isActive = React.useCallback(
    (val: string | number) => {
      if (innerValue === undefined) {
        return false
      }

      if (isMultiple && Array.isArray(innerValue)) {
        return innerValue.includes(val)
      }

      return innerValue === val
    },
    [innerValue]
  )

  const getFocusedValue = React.useCallback(() => {
    const item = activeOptions[focusedIndex]

    if (item) {
      return item.value
    }

    return null
  }, [activeOptions, focusedIndex])

  const sortOptions = React.useCallback(() => {
    if (isNull(innerValue) || !options) return

    let sortedOptions
    if (Array.isArray(innerValue)) {
      const selectedOptions = innerValue
        .map(value => options.find(option => option.value === value))
        .filter(option => option !== undefined)
      sortedOptions = Array.from(new Set([...selectedOptions, ...options]))
    } else {
      sortedOptions = [...options].sort(item => {
        if (item.value !== innerValue) return 1

        return -1
      })
    }

    setInnerOptions(sortedOptions as SelectOption[])
  }, [innerValue, options])

  const setValue = (value: string | number | undefined): void => {
    if (isMultiple && value !== undefined) {
      const pos = ((innerValue as Array<string | number>) || []).indexOf(value)
      const data = [...((innerValue as Array<string | number>) || [])]

      if (pos >= 0) {
        data.splice(pos, 1)
      } else {
        data.push(value)
      }

      setInnerValue(data, val => {
        //TODO проверить типизацию в useStateCallback
        safeInvoke(onChange, val as SelectValue)
      })

      return
    }

    setInnerValue(value, val => {
      //TODO проверить типизацию в useStateCallback
      safeInvoke(onChange, val as SelectValue)
    })
  }

  const hideMenu = () => {
    setVisible.off()
    inputRef.current?.focus()
  }

  const handleInputBlur = React.useCallback(
    event => {
      !focused && setVisible.off()
      !focused && safeInvoke(onBlur, event)
    },
    [focused, onBlur]
  )

  const handleInputClick = React.useCallback(
    (event: React.MouseEvent<HTMLInputElement>) => {
      !readOnly && setVisible.toggle()
      safeInvoke(onClick, event)
    },
    [readOnly, onClick]
  )

  const handleInputFocus = React.useCallback(
    (event: React.FocusEvent<HTMLInputElement>) => {
      !focused && setFocused.on()
      !focused && safeInvoke(onFocus, event)
    },
    [focused, onFocus]
  )

  const handleSelect = (value: string | number) => (_event: React.MouseEvent) => {
    setValue(value)

    isMultiple ? inputRef.current?.focus() : hideMenu()
  }

  const handleMouseMove = () => {
    setFocusedIndex(getCurrentIndex(innerValue))
  }

  const menu = React.useMemo(() => {
    if (!innerOptions) {
      return null
    }

    return (
      <Menu>
        {innerOptions.map(({ label, value, disabled, ...args }) => (
          <Menu.Item
            key={value}
            ref={getFocusedValue() === value ? focusedItemRef : null}
            {...args}
            value={value}
            active={isActive(value) || getFocusedValue() === value}
            disabled={disabled}
            onClick={disabled ? noop : handleSelect(value)}
            onMouseOver={handleMouseMove}
          >
            {label}
          </Menu.Item>
        ))}
      </Menu>
    )
  }, [innerOptions, innerValue, focusedIndex, getFocusedValue, isActive])

  const view = React.useMemo(() => {
    return (
      <Portal>
        <Fade
          ref={popperRef}
          className={cx('menu')}
          unmountOnExit
          visible={visible}
          {...getPopperProps({ style: { zIndex: Z_INDEXES.DROPDOWN } }, popper)}
        >
          {menu}
        </Fade>
      </Portal>
    )
  }, [visible, menu])

  const handleCrossIconClick = () => {
    setFocusedIndex(-1)
    setValue(undefined)
  }

  const handleArrowClick = (event: React.MouseEvent) => {
    if (visible) {
      event.stopPropagation()
      hideMenu()
    }
  }

  const iconElement = React.useMemo(() => {
    if (icon) {
      return icon
    }

    if (withCrossIcon && visible && innerValue !== undefined) {
      return <Cross className={cx('icon')} onClick={handleCrossIconClick} />
    }

    return <ArrowDown className={cx('icon', { [`isArrowTurned`]: visible })} onClick={handleArrowClick} />
  }, [visible, innerValue])

  useKeyPress(
    {
      enabled: visible || focused,
      keyMap: {
        ArrowDown: event => {
          event.preventDefault()
          if (!visible) {
            setVisible.on()
            return
          }
          const nextIndex = getNextIndex(focusedIndex, activeOptions.length)
          setFocusedIndex(nextIndex)
        },

        ArrowUp: event => {
          event.preventDefault()
          if (!visible) {
            setVisible.on()
            return
          }
          const prevIndex = getPrevIndex(focusedIndex, activeOptions.length)
          setFocusedIndex(prevIndex)
        },

        Enter: event => {
          event.preventDefault()
          if (!visible) {
            setVisible.on()
            return
          }
          const value = getFocusedValue()

          if (value !== undefined && value !== null) {
            setValue(value)

            if (!isMultiple) {
              hideMenu()
            }
          }
        },

        Escape: event => {
          event.preventDefault()
          hideMenu()
        },

        Tab: () => {
          setVisible.off()
          setFocused.off()
        },
      },
    },
    [innerValue, focusedIndex, activeOptions, visible, focused]
  )

  React.useEffect(() => {
    if (visible) {
      setFocusedIndex(getCurrentIndex(innerValue))
      setFocused.on()
    }
  }, [visible, innerValue])

  React.useEffect(() => {
    const listener = (event: Event) => {
      // @ts-ignore
      if (reference.current && reference.current.contains(event.target)) {
        return
      }

      // @ts-ignore
      if (popper.current && popper.current.contains(event.target)) {
        return
      }

      visible && hideMenu()
      setFocused.off()
    }

    document.addEventListener('mousedown', listener)

    return () => {
      document.removeEventListener('mousedown', listener)
    }
  }, [visible, reference, popper])

  React.useEffect(() => {
    if (value !== undefined) {
      setInnerValue(value)
    }
  }, [value])

  React.useEffect(() => {
    !visible && sortOptions()
  }, [visible, options, sortOptions])

  return (
    <div data-qa="Select" className={classNames}>
      <div ref={referenceRef} {...getReferenceProps({}, reference)} onClick={handleInputClick}>
        <TextInput
          {...restProps}
          autoFocus={autoFocus}
          ref={useMergedRef(ref, inputRef)}
          invalid={formControlContext.invalid || invalid}
          disabled={formControlContext.disabled || disabled}
          value={displayValue}
          icon={iconElement}
          readOnly={true}
          onFocus={handleInputFocus}
          onBlur={handleInputBlur}
        />
      </div>
      {view}
    </div>
  )
})

Select.defaultProps = {
  withCrossIcon: true,
  fixedWidth: false,
  placement: 'bottom-start',
  flip: true,
}

export default Select
