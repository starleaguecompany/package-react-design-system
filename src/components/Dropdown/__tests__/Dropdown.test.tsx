import * as React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Dropdown, Button } from '../../..';
import { SelectOption } from '../../../types/Select.types';

const options: SelectOption[] = [
  { label: 'label 1', value: 1 },
  { label: 'label 2', value: 2 },
  { label: 'label 3', value: 3 },
];

const child = <Button>OK</Button>;

describe('Dropdown', () => {
  test('should render correctly with no props', () => {
    render(<Dropdown>{child}</Dropdown>);

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    expect(screen.getByTestId('Dropdown')).toBeVisible();
  });

  test('should render correctly with attributes', () => {
    render(
      <Dropdown id="test-id" className="test-class">
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    expect(screen.getByTestId('Dropdown')).toHaveClass('test-class');
    expect(screen.getByTestId('Dropdown')).toHaveAttribute('id', 'test-id');
  });

  test('should render correctly with required option prop', () => {
    render(<Dropdown options={options}>{child}</Dropdown>);

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    expect(screen.getByRole('button')).toBeVisible();
    expect(screen.queryByText('label 1')).not.toBeInTheDocument();
    expect(screen.queryByText('label 2')).not.toBeInTheDocument();
    expect(screen.queryByText('label 3')).not.toBeInTheDocument();
  });

  test('should display list of option by click on dropdown element', async () => {
    render(<Dropdown options={options}>{child}</Dropdown>);

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    userEvent.click(screen.getByRole('button'));
    await waitFor(() => expect(screen.getByText('label 1')).toBeVisible());
    await waitFor(() => expect(screen.getByText('label 2')).toBeVisible());
    await waitFor(() => expect(screen.getByText('label 3')).toBeVisible());
  });

  test('should hide list of option by click outside', async () => {
    render(<Dropdown options={options}>{child}</Dropdown>);

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    userEvent.click(screen.getByRole('button'));
    userEvent.click(document.body);
    await waitFor(() => expect(screen.queryByText('label 1')).not.toBeInTheDocument());
    await waitFor(() => expect(screen.queryByText('label 2')).not.toBeInTheDocument());
    await waitFor(() => expect(screen.queryByText('label 3')).not.toBeInTheDocument());
  });

  test('should change active item in list of option on select it', () => {
    render(
      <Dropdown value={1} options={options}>
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    userEvent.click(screen.getByRole('button'));
    userEvent.click(screen.getByText('label 3'));
    userEvent.click(screen.getByRole('button'));
    expect(screen.getByText('label 1').closest('.itemBase')).not.toHaveClass('active');
    expect(screen.getByText('label 2').closest('.itemBase')).not.toHaveClass('active');
    expect(screen.getByText('label 3').closest('.itemBase')).toHaveClass('active');
  });

  test('should render correctly with defaultValue prop', () => {
    render(
      <Dropdown defaultValue={2} options={options}>
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    userEvent.click(screen.getByRole('button'));
    expect(screen.getByText('label 1').closest('.itemBase')).not.toHaveClass('active');
    expect(screen.getByText('label 2').closest('.itemBase')).toHaveClass('active');
    expect(screen.getByText('label 3').closest('.itemBase')).not.toHaveClass('active');
  });

  test('should render correctly with value prop', () => {
    render(
      <Dropdown value={3} options={options}>
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    userEvent.click(screen.getByRole('button'));
    expect(screen.getByText('label 1').closest('.itemBase')).not.toHaveClass('active');
    expect(screen.getByText('label 2').closest('.itemBase')).not.toHaveClass('active');
    expect(screen.getByText('label 3').closest('.itemBase')).toHaveClass('active');
  });

  test('should render correctly with placement prop', () => {
    render(
      <Dropdown placement="left" options={options}>
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    userEvent.click(screen.getByRole('button'));
    const dropdownContainer = document.querySelector('[data-popper-placement="left"]');
    expect(dropdownContainer).toBeInTheDocument();
  });

  test('should execute onChangeCallback on select item', () => {
    const onChangeCallback = jest.fn(value => value);
    render(
      <Dropdown onChange={onChangeCallback} options={options}>
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    userEvent.click(screen.getByRole('button'));
    expect(onChangeCallback).toBeCalledTimes(0);
    userEvent.click(screen.getByText('label 3'));
    expect(onChangeCallback).toBeCalledTimes(1);
    expect(onChangeCallback).toReturnWith(3);
  });

  test('should return selected value on select item', () => {
    let value;
    render(
      <Dropdown onChange={e => (value = e)} options={options}>
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    userEvent.click(screen.getByRole('button'));
    userEvent.click(screen.getByText('label 3'));
    expect(value).toEqual(3);
  });

  test('should execute callbacks on open and on close dropdown', () => {
    const onOpenCallback = jest.fn();
    const onCloseCallback = jest.fn();
    render(
      <Dropdown onOpen={onOpenCallback} onClose={onCloseCallback} options={options}>
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    expect(onOpenCallback).toBeCalledTimes(0);
    expect(onCloseCallback).toBeCalledTimes(0);
    userEvent.click(screen.getByRole('button'));
    expect(onOpenCallback).toBeCalledTimes(1);
    expect(onCloseCallback).toBeCalledTimes(0);
    userEvent.click(document.body);
    expect(onCloseCallback).toBeCalledTimes(1);
  });

  test('should render correctly with mode prop', () => {
    render(
      <Dropdown options={options} mode="multiple">
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();
    userEvent.click(screen.getByRole('button'));
    userEvent.click(screen.getByText('label 2'));
    userEvent.click(screen.getByText('label 3'));
    expect(screen.getByText('label 1').closest('.itemBase')).not.toHaveClass('active');
    expect(screen.getByText('label 2').closest('.itemBase')).toHaveClass('active');
    expect(screen.getByText('label 3').closest('.itemBase')).toHaveClass('active');
  });

  test('should render correctly on external value update', () => {
    const onChangeCallback = jest.fn();
    const { rerender } = render(
      <Dropdown options={options} value={1} onChange={onChangeCallback}>
        {child}
      </Dropdown>
    );

    expect(screen.getByTestId('Dropdown')).toMatchSnapshot();

    userEvent.click(screen.getByRole('button'));
    expect(screen.getByText('label 1').closest('.itemBase')).toHaveClass('active');
    expect(screen.getByText('label 2').closest('.itemBase')).not.toHaveClass('active');
    expect(screen.getByText('label 3').closest('.itemBase')).not.toHaveClass('active');

    rerender(
      <Dropdown options={options} value={2} onChange={onChangeCallback}>
        {child}
      </Dropdown>
    );

    userEvent.click(screen.getByRole('button'));
    expect(screen.getByText('label 1').closest('.itemBase')).not.toHaveClass('active');
    expect(screen.getByText('label 2').closest('.itemBase')).toHaveClass('active');
    expect(screen.getByText('label 3').closest('.itemBase')).not.toHaveClass('active');
    expect(onChangeCallback).toBeCalledTimes(0);
  });
});
