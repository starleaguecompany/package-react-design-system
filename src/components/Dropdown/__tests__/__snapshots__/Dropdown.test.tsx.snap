// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should change active item in list of option on select it 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should display list of option by click on dropdown element 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should execute callbacks on open and on close dropdown 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should execute onChangeCallback on select item 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should hide list of option by click outside 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should render correctly on external value update 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should render correctly with attributes 1`] = `
<button
  class="container test-class container size-44 color-gray-outlined"
  data-qa="Dropdown"
  id="test-id"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should render correctly with defaultValue prop 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should render correctly with mode prop 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should render correctly with no props 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should render correctly with placement prop 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should render correctly with required option prop 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should render correctly with value prop 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;

exports[`Dropdown should return selected value on select item 1`] = `
<button
  class="container container size-44 color-gray-outlined"
  data-qa="Dropdown"
  tabindex="0"
>
  <span
    class="text"
  >
    OK
  </span>
</button>
`;
